src/dmd/expression.d

```diff
@@ -5307,6 +5307,49 @@ extern (C++) final class DotVarExp : UnaExp
         if (e1.op == TOK.this_)
             return var.checkModify(loc, sc, e1, flag);
 
+        // Issue 12764 : https://issues.dlang.org/show_bug.cgi?id=12764
+
+        // If inside a constructor and an expression of type this.field.var
+        // is encountered, where field is a struct declaration with
+        // default construction disabled, we must check if that field was
+        // properly initialized before being able to actually use its
+        // fields.
+        if (sc.func)
+        {
+            auto ctd = sc.func.isCtorDeclaration();
+
+            // if inside a constructor scope and e1 of this DotVarExp
+            // is a DotVarExp, then check if e1.e1 is a `this` identifier
+            if (ctd && e1.op == TOK.dotVariable)
+            {
+                scope dve = cast(DotVarExp)e1;
+                if (dve.e1.op == TOK.this_)
+                {
+                    scope v = dve.var.isVarDeclaration();
+                    // if v is a struct member field with no initializer and no default construction
+                    if (v && v.isField() && v.type.ty == Tstruct && !v._init)
+                    {
+                        const sd = (cast(TypeStruct)v.type).sym;
+                        if (sd.noDefaultCtor)
+                        {
+                            // check if v was initialized. if not we must error
+                            const ad = ctd.isMember2();
+                            size_t i;
+                            for (i = 0; i < sc.ctorflow.fieldinit.length; i++)
+                            {
+                                if (ad.fields[i] == v)
+                                    break;
+                            }
+                            const fi = sc.ctorflow.fieldinit[i];
+                            import dmd.ctorflow : CSX;
+                            if (!(fi & CSX.this_ctor))
+                                .error(loc, "`struct` field `%s` must be initialized before accessing any of its members as it cannot be default constructed", v.toChars());
```

Comment:
The bug is not that it must be initialized, but that it must be initialized *with a constructor call*.

Reply:
Also, the bug is about `s.f = 1;` which is initialization of `s` by assignment. But this patch is also detecting rvalue usage, such as `x = s.f;`, which should be allowed.

Reply:
>Also, the bug is about s.f = 1; which is initialization of s by assignment. But this patch is also detecting rvalue usage, such as x = s.f;, which should be allowed.
 

That is not true. The current patch does not affect the case of `x = s.f;` because checkModifiable is called inside the AssignExp semantic visitor node for `lhs` of the AssignExp. checkModifiable is never called for a `rhs`.

>The bug is not that it must be initialized, but that it must be initialized with a constructor call.

It looks like you are right, although I find it harder and harder to understand why it should be allowed to change the value of `s.f` (which is equivalent to partially initializing the struct) **before** calling the constructor